frontend/src/components/applyForm/utils.tsx

```diff
@@ -0,0 +1,197 @@
+import { RJSFSchema } from "@rjsf/utils";
+import { ErrorObject } from "ajv";
+import { get as getSchemaObjectFromPointer } from "json-pointer";
+import { filter, get } from "lodash";
+
+import { JSX } from "react";
+
+import { UiSchema } from "./types";
+import { FieldsetWidget } from "./widgets/FieldsetWidget";
+import SelectWidget from "./widgets/SelectWidget";
+import TextareaWidget from "./widgets/TextAreaWidget";
+import TextWidget from "./widgets/TextWidget";
+
+export function buildFormTreeClosure(
+  schema: RJSFSchema,
+  uiSchema: UiSchema,
+  errors: ErrorObject<string, Record<string, unknown>, unknown>[],
+  formData: object,
+) {
+  let acc: JSX.Element[] = [];
+
+  const buildFormTree = (
+    uiSchema: UiSchema,
+    parent: { label: string; name: string } | null,
+  ) => {
+    if (
+      !Array.isArray(uiSchema) &&
+      typeof uiSchema === "object" &&
+      "children" in uiSchema
+    ) {
+      buildFormTree(uiSchema.children, {
+        label: uiSchema.label,
+        name: uiSchema.name,
+      });
+    } else if (Array.isArray(uiSchema)) {
+      uiSchema.forEach((node) => {
+        if ("children" in node) {
+          buildFormTree(node.children, {
+            label: node.label,
+            name: node.name,
+          });
+        } else if (!parent && "definition" in node) {
+          const field = buildField(node.definition, schema, errors, formData);
+          if (field) {
+            acc = [...acc, field];
+          }
+        }
+      });
+      if (parent) {
+        // eslint-disable-next-line array-callback-return
+        const row = uiSchema.map((node) => {
+          if ("children" in node) {
+            // TODO: remove children from acc
+            // return children from acc;
+          } else {
+            const { definition } = node as { definition: string };
+            return buildField(definition, schema, errors, formData);
+          }
+        });
+        acc = [...acc, wrapSection(parent.label, parent.name, <>{row}</>)];
+        // acc = parentHasChild(uiSchema) ? wrapSection(parent.label, parent.name, row) : [acc, wrapSection(parent.label, parent.name, row)];
+      }
+    }
+  };
+  buildFormTree(uiSchema, null);
+  return acc;
+}
+
+const createField = (
+  id: string,
```

Comment:
wow that's a lot of arguments. I'd say this is non-blocking but it's more annoying to make this sort of change later on - what do you say to packing this into an object argument so that we don't have to worry about argument order?

Reply:
Yeah, I meant to remove this function. I updated it. Thanks for noting it.